### PR TITLE
[OptApp] Integration of NLOPT libraray

### PR DESCRIPTION
**Pull request by RezaNajian**
This PR introduces an interface to the NLOPT optimizers. Currently, it supports three gradient-based optimizers: MMA, SLSQP, and LBFGS. The interface is designed in a flexible manner to ensure gradient-free optimizers can also be integrated with ease. However, our current focus is on gradient-based methods.

The interface is implemented at the Python level, and the NLOPT Python library needs to be in the Python path.

A shell thickness optimization case has been added and tested using the MMA algorithm.

### CODE REVIEW

File: applications/OptimizationApplication/tests/NLOPT_tests/MMA_shell_thickness_opt/test_NLOP_optimizers.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+try:
+    import nlopt
+    nlopt_available = True
+except ImportError:
+    nlopt_available = False
+
+@kratos_unittest.skipIf(not nlopt_available, "Missing nlopt python libraries ")
+class TestNLOPTOptimizers(kratos_unittest.TestCase):
+    # @classmethod
+    # def setUpClass(cls):
```
Comment (Igarizza) [inline applications/OptimizationApplication/tests/NLOPT_tests/MMA_shell_thickness_opt/test_NLOP_optimizers.py:17]: left comments

File: applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        for constraint_settings in parameters["constraints"]:
+            self.__constraints.append(StandardizedNLOPTConstraint(constraint_settings, self.master_control, self._optimization_problem))
+
+        # nlopt settings
+        NLOPT_settings = parameters["NLOPT_settings"]
+        NLOPT_settings.ValidateAndAssignDefaults(self.GetDefaultParameters()["NLOPT_settings"])
+
+        # nlopt algorithm
+        self.algorithm_name = NLOPT_settings["algorithm_name"].GetString()
+        if(len(self.__constraints)==0):
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py:72]: can you update the CheckOptimizerSupport to take _optimization_problem and do all the checks inside to keep __init__ cleaner?

File: applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+            "constraints"         : [],
+            "controls"          : [],
+            "echo_level"        : 0,
+            "NLOPT_settings"          : {
+                "algorithm_name"      : "MMA",
+                "stopping_criteria"   : {
+                    "objective_rel_tol": 0,
+                    "objective_abs_tol": 0,
+                    "controls_rel_tol": 0,
+                    "controls_abs_tol": 0,
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py:40]: to be consistent with used names in steepest descent:

rel_obj_tol

abs_obj_tol

rel_contr_tol

abs_contr_tol // contr hasn't been used before, you can use any prefered short version.

File: applications/OptimizationApplication/tests/NLOPT_tests/MMA_shell_thickness_opt/optimization_parameters.json [hunk 4]
```diff
@@ -0,0 +1,124 @@
+{
```
Comment (Igarizza) [inline applications/OptimizationApplication/tests/NLOPT_tests/MMA_shell_thickness_opt/optimization_parameters.json:1]: please move NLOPT_tests folder into algorithm_test.

File: applications/OptimizationApplication/python_scripts/controls/master_control.py [hunk 5]
```diff
@@ -162,23 +163,34 @@ def MapGradient(self, physical_space_gradient_variable_and_collective_expression
...
 
         Raises:
             RuntimeError: If number of controls and number of container expressions mismatch.
 
         Returns:
             dict[Control, bool]: A map with control and a boolean whether the update changed anything in that control.
         """
-        if len(self.__list_of_controls) != len(update_collective_expressions.GetContainerExpressions()):
+
+        if isinstance(updated_control_vars, numpy.ndarray):
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/controls/master_control.py:179]: I am not sure in this implementation. Why not to convert numpy back to container expression and keep standart well tested routine?

File: applications/OptimizationApplication/python_scripts/responses/response_routine.py [hunk 6]
```diff
@@ -70,7 +70,7 @@ def Finalize(self):
     def GetReponse(self) -> ResponseFunction:
         return self.__response
 
-    def CalculateValue(self, control_field: KratosOA.CollectiveExpression) -> float:
+    def CalculateValue(self, control_field: KratosOA.CollectiveExpression=None) -> float:
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/responses/response_routine.py:73]: See sugestion in the master control.

File: applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py [hunk 7]
```diff
@@ -203,7 +203,7 @@ def Update(self, new_control_field: ContainerExpressionTypes) -> bool:
         if not IsSameContainerExpression(new_control_field, self.GetEmptyField()):
             raise RuntimeError(f"Updates for the required element container not found for control \"{self.GetName()}\". [ required model part name: {self.model_part.FullName()}, given model part name: {new_control_field.GetModelPart().FullName()} ]")
 
-        if KratosOA.ExpressionUtils.NormL2(self.control_field - new_control_field) > 1e-9:
+        if KratosOA.ExpressionUtils.NormL2(self.control_field - new_control_field) > 1e-15:
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py:206]: Should we move this tolerance into JSON file to allow changing it?

File: applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_objective.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+class StandardizedNLOPTObjective(ResponseRoutine):
+    """Standardized objective response function
+
+    This class transformed a user-given optimization problem into the standard format.
+    Supported objective types:
+        "minimization",
+        "maximization"
+
+    """
+    def __init__(self, parameters: Kratos.Parameters, master_control: MasterControl, optimization_problem: OptimizationProblem, required_buffer_size: int = 2):
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_objective.py:23]: Why not to inherit from standartized objective.py and reloud only few functions. A lot of functions are identical.

File: applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_objective.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+            return self.__unbuffered_data["initial_value"] * self.__scaling
+        else:
+            raise RuntimeError(f"Response value for {self.GetReponse().GetName()} is not calculated yet.")
+
+    def Initialize(self):
+        super().Initialize()
+
+    def CalculateStandardizedValueAndGradients(self, control_field: numpy.ndarray, gradient_field: numpy.ndarray, save_value: bool = True) -> float:
+
+        # update the master control
```
Comment (Igarizza) [inline applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_objective.py:68]: Can you please comment whtat is happening here and why do you need it?

**Review by Igarizza — COMMENTED** [review]
@RezaNajian I am happy to see that your and @sunethwarna design allow this clean implementation of the external opt. library. In general, the PR is nice and clean. Still, I think there are places for further improvement.

File: applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py (hunk 2 above)
Comment (philbucher) [inline applications/OptimizationApplication/python_scripts/algorithms/NLOPT_algorithms.py:72]:
```suggestion
        if not self.__constraints:
```

File: applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_objective.py (hunk 9 above)
Comment (RezaNajian) [reply applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_objective.py:68]: @Igarizza I added more comments to the function, have a look to the new PR #11480

File: applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_objective.py (hunk 8 above)
Comment (RezaNajian) [reply applications/OptimizationApplication/python_scripts/algorithms/standardized_NLOPT_objective.py:23]: I have overwritten almost function only GetValue and GetStandardizedValue are the same, for the sake of clarity let's  keep it fully deparate

File: applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py (hunk 7 above)
Comment (RezaNajian) [reply applications/OptimizationApplication/python_scripts/controls/thickness/shell_thickness_control.py:206]: I do not know honestly, I am open to suggestions

File: applications/OptimizationApplication/python_scripts/controls/master_control.py (hunk 5 above)
Comment (RezaNajian) [reply applications/OptimizationApplication/python_scripts/controls/master_control.py:179]: @Igarizza that's a very good point. Applied !

File: applications/OptimizationApplication/python_scripts/responses/response_routine.py (hunk 6 above)
Comment (RezaNajian) [reply applications/OptimizationApplication/python_scripts/responses/response_routine.py:73]: also applied. Tnx.

**Comment by RezaNajian** [comment]
this PR and the corresponding reviews are moved and addressed in PR #11480.

**Pull request merged**